Refactor how modules are declared and created

blinkseb opened the pull request:
First of all, this PR is big, sorry in advance for the reviewers! It's also highly technical and I'm pretty sure I won't be able to explain everything in this description. That's why I'd propose to discuss more IRL Friday after the meeting so I can better explain the whys.

---

In the current implementation, modules are ordered using a [directed acyclic graph (DAG)](https://en.wikipedia.org/wiki/Directed_acyclic_graph), built based on modules' inputs and outputs. The shortcoming here is that we *do not* know theses information before creating the modules (there's actually a spy in the memory pool which tracks the inputs and outputs allocations). This kind of works, but it has a some drawbacks. Here's some I can think of right now:
  - there's no way to validate the configuration file since we don't know what the inputs or outputs are. Currently we just throw an exception indicating that a module requested a non-existent parameter / input
   - #78, unfixable in the current implementation (I can explain why IRL if you want)
   - Only the documentation actually explain what inputs / parameters are needed by the module.

I did some work to rework how we declare the modules in order to now specify what are the inputs, outputs and parameters (= attributes) needed by a module in order to work. The way it's done is copied directly from TensorFlow, as well as some bit of codes (namely the `StringPiece`, `optional` and `ModuleBuilder` stuff). This is reflected in the copyright notice of the concerned classes hopefully.

You have now a bit more work to do when registering a module into the system. For example, let's look at the `MatrixElement` module:

```C++
REGISTER_MODULE(MatrixElement)
        .Input("initialState")
        .OptionalInputs("jacobians")
        .Inputs("particles/inputs")
        .Output("output")
        .GlobalAttr("energy:double")
        .Attr("matrix_element:string")
        .Attr("matrix_element_parameters:pset")
        .OptionalAttr("override_parameters:pset")
        .Attr("particles:pset")
        .Attr("use_pdf:bool=true")
        .OptionalAttr("pdf:string")
        .OptionalAttr("pdf_scale:double");
```

This (I hope) should be straightforward to understand. You register the module, and then, with chained calls, define what are needed for this module. Requirements are split into three categories:
  - Input(s): those are the InputTag(s) needed by the modules
  - Output: what is produced by this module
  - Attr (attributes) : the parameters needed. You have to specify the type (list of supported types available in the documentation) and possibly a default value. These are not used for the moment but will be later to further validate the configuration file and provide insightful warnings and comments.
Inputs and Attr can be optional. In this case, no warning is shown is the values are not specified in the configuration file. 

Basic validation of the configuration file is now performed before creating the modules. By validation, I mean:
   - any mandatory inputs or attribute missing from the configuration file result in an error
   - a warning is shown for any unused parameter in the configuration (useful to spot spelling error for example, if the user passed `input = blabla` instead of `inputs = blabla`, a message telling that `input` is not used by the module will be shown)

I also took the opportunity to improve how the DAG is generated, mainly to be able to fix #78. A new class is introduced, `ComputationGraph`, embedding the DAG (which is, in its final form, only an ordered list of modules for the moment). Since we can now build the DAG without creating any modules, this allow us to change or edit the parameters before creating the modules. It's useful to solve #78: imagine there're 4 modules into the graph, each requesting an integration dimension. They'll have for InputTag `cuba::ps_point/1`, ..., `cuba::ps_point/4`. Now, imagine the second module is not used anywhere in the computation. The DAG will automatically remove it, but the number of dimension is still 4. You can recompute it "easily" and you see there's only 3, great! But the InputTag for the other modules are still `cuba::ps_point/1`, `cuba::ps_point/3` and `cuba::ps_point/4`, which is very bad since you only have 3 dimensions (1, 2 and 3). To fix this, there's now a re-indexing of all the cuba InputTags before creating the module, ensuring a continuous indexing if the number of dimension changes after the DAG optimization :tada:. As a bonus, how the Loopers work is now cleaner :smile: 

I also added unit tests for the DAG and other parts of the code, which should cover all the weird cases we should support.

I've put the milestone to v1.0, but this need to be discussed. I don't wan't to hold back the release because we are reviewing this!

swertz commented on include/momemta/ModuleDefBuilder.h:
```diff
@@ -0,0 +1,147 @@
+namespace momemta {
+
+/**
+ * A container for module registration data
```
What's the need for this extra layer?

blinkseb replied:
It used to contains more than the module definition, but indeed now I don't see a real usage

blinkseb replied:
I prefer to keep it, it may be useful in the future if we want to extend the module registration

swertz commented on include/momemta/Pool.h:
```diff
@@ -148,21 +111,13 @@ class Pool {
         Pool& operator=(const Pool&) = delete;
 
-        Configuration::Module m_current_module; /// Module currently created.
+        Configuration::ModuleDecl m_current_module; /// Module currently created.
```
That's not used anymore, right?

blinkseb replied:
Indeed, nice catch!

swertz commented on include/momemta/ModuleDef.h:
```diff
@@ -0,0 +1,77 @@
+struct AttrDef {
+    std::string name;
+    std::string type;
+    std::string default_value;
```
Are the default values defined here used or checked anywhere? What's really used is the default value defined by the user in the module constructor, or did I miss something?

blinkseb replied:
For the moment, it's not used anywhere you're right. There's some complication here because you have to convert from string to the proper type, and that's not straightforward. The long-term goal here is to directly inject the parameter with its default value in the `ParameterSet` if it's not specified by the user, which would mean that the `default` value in the `get` method of pset would non longer be relevant. But for the moment, see that more as a way to quickly see if some attributes of the modules have default values than anything else.

blinkseb commented on core/src/Graph.cc:
```diff
@@ -50,157 +87,313 @@ bool isConnectedToByIn(Graph& g, vertex_t vertex, vertex_t to) {
+        for (const auto& output: vertex.def.outputs) {
+
+            // Find any module using this output
+            for (const auto& module: requested_modules) {
```
IIRC, it was because first the definition was not stored in the vertex, so it was the only way. I'll change to iterate over the vertices instead, much cleaner indeed!

blinkseb replied:
Done :+1:

blinkseb commented on core/src/Graph.cc:
```diff
@@ -50,157 +87,313 @@ bool isConnectedToByIn(Graph& g, vertex_t vertex, vertex_t to) {
     // is ran before the looper itself.
     for (const auto& vertex: vertices) {
-        if (g[vertex.second].configuration_module.type == "Looper") {
+        if (g[vertex.second].type == "Looper") {
```
Done, good suggestion :+1:

blinkseb commented on core/src/Graph.cc:
```diff
@@ -221,66 +414,201 @@ Graph build(const Pool::DescriptionMap& description, std::vector<ModulePtr>& mod
+            }
+
+            const auto& inputTag = g[*o].tag;
+            assert(inputTag.isIndexed());
```
Hm, I'm not sure why this code is here :sweat_smile: Looks like a forgotten copy-paste or the like. Thanks for catching, I'll remove it

blinkseb commented on core/src/Graph.cc:
```diff
@@ -221,66 +414,201 @@ Graph build(const Pool::DescriptionMap& description, std::vector<ModulePtr>& mod
             }
-        }
+
+            // Check in the definition if this module has any output
```
Not sure if I understand your question correctly. I made the assumption here that a module without output is necessarily sticky as it would not make sense otherwise. IIUC, you propose enforce the use of the `sticky` attribute no matter if a module has output or not. It's fine for me, it's maybe more consistent than trying to guess.

swertz replied:
Yes that was what I meant ;) I don't really mind either way as long as it's documented (it's not what is written here at the moment: https://github.com/MoMEMta/momemta.github.io/pull/14/files#diff-1f46ac87a66f014d74627b97bcb7caa9R49)

blinkseb replied:
Fair enough, I'll remove the check and force the module to be sticky 👍 Maybe log a warning?

swertz replied:
Fine with me ;)

blinkseb commented on core/src/Graph.cc:
```diff
@@ -318,36 +646,184 @@ Graph build(const Pool::DescriptionMap& description, std::vector<ModulePtr>& mod
         std::string one_of_the = it->second.size() ? "one of the" : "the";
```
Probably yes :smile:

blinkseb commented on core/src/registry/ModuleRegistry.cc:
```diff
@@ -0,0 +1,129 @@
+void ModuleRegistry::registerModule(RegisterOp registration_op) {
+    std::unique_lock<std::mutex> lock(mutex_);
+    if (!initialized_) {
+        deferred_.emplace_back(registration_op);
```
This part is stolen from TensorFlow. I imagine they need this to ensure some other subsystems are initialized before registering the module. In our case for the moment it's not really useful, but can be in the future...

blinkseb commented on core/src/registry/ModuleDefBuilder.cc:
```diff
@@ -0,0 +1,259 @@
+    def.many = spec.Consume("*");
+
+    if (! is_output) {
+        std::vector<std::string> nested_attributes;
```
Indeed. And there's a FIXME right below I missed!

blinkseb commented on core/src/registry/ModuleDefBuilder.cc:
```diff
@@ -0,0 +1,259 @@
+
+    // An optional input starts with a '?'
+
+    def.optional = spec.Consume("?");
```
IMO it's much easier to use than regex, and probably faster too. One advantage is that the `Scanner` works with `StringPiece` while regex does not. This avoids a lot of string copy

blinkseb commented on core/src/Graph.cc:
```diff
@@ -50,157 +87,313 @@ bool isConnectedToByIn(Graph& g, vertex_t vertex, vertex_t to) {
-    // grab its execution path from its configuration
-    const auto& looper_path = g[looper].configuration_module.parameters->get<Path>("path");
-    const auto& path_modules = looper_path.modules();
+momemta::ModuleList::value_type get_module_def(const std::string& module_type,
```
It's the only place it's used, I'm not sure where is the right place to locate it

swertz replied:
Ok since it's the only place it's probably fine!

blinkseb commented:
I've pushed some commits addressing the comments, thanks a lot!

Concerning the place where the configuration should be validated, that's a good question. My worry was the fact that the user can still update and modify the configuration file after it has been parsed. Maybe inside `freeze` is the best location?

swertz commented:
:+1: I still see a couple of FIXMEs but they're probably not urgent... I'm fine with leaving them for later, just tell me if you plan on adding more stuff before merging this ;)

Yes that was my suggestion, doing it when the `Configuration` itself is created (but as you say it needs to be done inside the `ConfigurationReader` class, ie the `freeze` method)... On the other hand I don't see why the user wouldn't do straight `MoMEMta weight(configuration.freeze());`, so it's really equivalent if we do it in there or in `MoMEMta` itself?

blinkseb commented on core/src/MoMEMta.cc:
```diff
@@ -24,113 +25,128 @@
-    // Next, retrieve all the input tags for the components of the integrand
-    m_pool->current_module("momemta");
+    std::string export_graph_as;
+    if (configuration.getGlobalParameters().existsAs<std::string>("export_graph_as"))
```
I updated my PR on the website to add this

blinkseb commented:
I'll have another looks at the FIXME!

swertz commented on core/src/Graph.cc:
```diff
@@ -280,7 +280,6 @@ void ComputationGraph::logTimings() const {
 #endif
 
 void ComputationGraph::setNDimensions(size_t n) {
-    assert(n >= 0);
```
Don't we want that? Unless you mean to get rid of `assert` and throw instead?

blinkseb commented:
`size_t` is unsigned for the assert is useless (thanks GCC for the warning
!)

Le ven. 30 juin 2017 à 17:41, Sébastien Wertz <notifications@github.com> a
écrit :

> *@swertz* commented on this pull request.
> ------------------------------
>
> In core/src/Graph.cc
> <https://github.com/MoMEMta/MoMEMta/pull/151#discussion_r125069296>:
>
> > @@ -280,7 +280,6 @@ void ComputationGraph::logTimings() const {
>  #endif
>
>  void ComputationGraph::setNDimensions(size_t n) {
> -    assert(n >= 0);
>
> Don't we want that? Unless you mean to get rid of assert and throw
> instead?
>
> —
> You are receiving this because you modified the open/close state.
>
>
> Reply to this email directly, view it on GitHub
> <https://github.com/MoMEMta/MoMEMta/pull/151#pullrequestreview-47425695>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAXk4nnd1oRPxFN2ofkazQtH7hhsjlOAks5sJRcOgaJpZM4N_0IH>
> .
>
-- 
Sébastien Brochet

blinkseb commented:
I've added some checks to ensure the attributes, inputs and outputs are actually correctly defined (it was not the case for all the secondary blocks :flushed:)

blinkseb commented:
I found some cases when testing this PR where invalid inputs where not correctly reported (the user sees an exception about invalid slot in the memory pool, not very useful to understand what's going on). This looks like a regression from the old behavior. I'm adding the `do not merge` label until I fix this

swertz commented:
In which cases have you observed this??

blinkseb commented:
Simply using an InputTag pointing to a non-existing module. It fails with the following exception:

```
184: [2017-07-25 21:56:18.302] [fatal] Memory block 'tt_looper_wrong::jacobian' is flagged as invalid. This should not happen. Please open a bug report at <>.
```

Not sure what it was before, but I remember something more user friendly...

blinkseb commented:
And especially not an error message blaming us for a user mistake :smile:

blinkseb commented:
Ok it seems I removed this block of code while refactoring everything: https://github.com/MoMEMta/MoMEMta/blob/d2953d1504ffecbdd7a1a3cb85b2e56530ce7231/core/src/Graph.cc#L171-L198

I'll add it back, it should solve this!

blinkseb commented:
Here we go!

blinkseb commented:
Rebased! On my side it's good to merge, I tested it extensively with the ttH stuff and it's working as expected!

swertz commented:
Awesome :+1:

I do see a few FIXMEs still lying around. I think one or two of them are quick to fix, others can perhaps be left for later, as you prefer?
https://github.com/MoMEMta/MoMEMta/pull/151/files#diff-57a27989d95956c8b94c36557dc4f602R493
https://github.com/MoMEMta/MoMEMta/pull/151/files#diff-57a27989d95956c8b94c36557dc4f602R878
https://github.com/MoMEMta/MoMEMta/pull/151/files#diff-1b3818a9d06b230f11f02d4e19ece9b5R173
https://github.com/MoMEMta/MoMEMta/pull/151/files#diff-b3935cbf75a1dcbb3c0e29872a3218a6R104

blinkseb commented:
Good catch thanks! One should be transformed into a `TODO`, others should be easily fixable. I'll do it tomorrow!

blinkseb commented:
Done, sorry for the delay!

swertz commented:
Great, a (final) thank you! :+1: Let's roll!
